smolyay/basis.py

```diff
@@ -403,29 +478,312 @@ def _derivative(self, x):
+            order of derivative. Default is 1.
+
+        Returns
+        -------
+        scalar or ndarray
+            the values of the basis functions.
+        """
+        new_X = self.scale_to_domain(numpy.asarray(X), domain)
+        y = numpy.zeros(
+            [len(self)] + list(new_X.shape),
+            dtype=complex if any(bf._is_complex for bf in self) else float,
+        )
+        for i in range(len(self)):
+            y[i, :] = self[i].derivative(new_X, n)
+        y *= ((self.domain[1] - self.domain[0]) / (domain[1] - domain[0])) ** n
+        return y
+
+
+class ChebyshevFirstKindBasisFunctionSet(BasisFunctionSet):
+    """Set of Chebyshev polynomials of the first kind.
+
+    Parameters
+    ----------
+    num_terms : int
+        the number of terms in the set.
+    """
+
+    def __init__(self, num_terms):
+        basis_functions = [ChebyshevFirstKind(f) for f in range(num_terms)]
+        super().__init__(basis_functions)
+
+
+class ChebyshevSecondKindBasisFunctionSet(BasisFunctionSet):
+    """Set of Chebyshev polynomials of the second kind.
+
+    Parameters
+    ----------
+    num_terms : int
+        the number of terms in the set.
+    """
+
+    def __init__(self, num_terms):
+        basis_functions = [ChebyshevSecondKind(f) for f in range(num_terms)]
+        super().__init__(basis_functions)
+
+
+class TrigonometricBasisFunctionSet(BasisFunctionSet):
+    """Set of Trigonmetric equations.
+
+    Parameters
+    ----------
+    num_terms : int
+        the number of terms in the set.
+    """
+
+    def __init__(self, num_terms):
+        index_trig = numpy.arange(num_terms, dtype=int)
+        frequencies = numpy.where(
+            index_trig % 2 == 1, (1 + index_trig) // 2, -index_trig // 2
+        )
+        basis_functions = [Trigonometric(f) for f in frequencies]
+        super().__init__(basis_functions)
+
+
+class NestedBasisFunctionSet(BasisFunctionSet):
+    """Set of nested basis functions and sample points.
+
+    Parameters
+    ----------
+    basis_functions : list
+        Basis functions in set.
+
+    num_per_level : list
+        number of unique functions per level.
+
+    Raises
+    ------
+    IndexError
+        number of basis function does not match functions in each level.
+    TypeError
+        Basis functions must have the same domain and typing.
+    """
+
+    def __init__(self, basis_functions, num_per_level):
+
+        # ensure parameters are lists
+        if isinstance(basis_functions, BasisFunction):
+            basis_functions = [basis_functions]
+        # check validity of basis functions
+        if numpy.sum(num_per_level) != len(basis_functions):
+            raise ValueError("Number of basis functions does not match level specification.")
+        super().__init__(basis_functions)
+        self._num_per_level = numpy.array(num_per_level, dtype=int)
+        self._start_level, self._end_level = _growth.get_level_start_and_end(num_per_level)
```

Comment:
You can use your numpy cast here:
```suggestion
        self._start_level, self._end_level = _growth.get_level_start_and_end(self._num_per_level)
```